### PR TITLE
[Autoscaling] Add CPU recommender

### PR DESCRIPTION
**Pull request by barkbay**
This PR adds a CPU recommender to the Autoscaling controller.

Fixes https://github.com/elastic/cloud-on-k8s/issues/5823

### CODE REVIEW

**Review by pebrc — APPROVED** [review]
Nice work! 👍 

In my testing, I think at least, I was not able to observe CPU based autoscaling but that might be due to the simulated load I was using that required more storage more than anything else.
 One thing I noticed is that the ML autoscaling I had in place went a bit crazy and requested 9 ML nodes with 8 CPU each  only to scale them down later (without me having created any ML models) I could not find a flaw in you implementation though and think this is purely due to the decision Elasticsearch made.

File: pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+) (Recommender, error) {
+	// Check if user expects the resource to be managed by the autoscaling controller
+	hasResourceRange := autoscalingSpec.CPURange != nil
+
+	// Did we get a resource requirement from Elasticsearch ?
+	hasRequirement := !autoscalingPolicyResult.RequiredCapacity.Node.Processors.IsEmpty() ||
+		!autoscalingPolicyResult.RequiredCapacity.Total.Processors.IsEmpty()
+
+	if hasRequirement && autoscalingSpec.CPURange == nil {
+		statusBuilder.ForPolicy(autoscalingSpec.Name).RecordEvent(status.CPURequired, "Min and max cpu must be specified")
```
Comment (thbkrkr) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go:110]:
```suggestion
		statusBuilder.ForPolicy(autoscalingSpec.Name).RecordEvent(status.CPURequired, "Min and max CPU must be specified")
```

File: pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+		return nil, fmt.Errorf("min and max cpu must be specified")
```
Comment (thbkrkr) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go:111]:
```suggestion
		return nil, fmt.Errorf("min and max CPU must be specified")
```

File: pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+func (m *cpu) ManagedResource() corev1.ResourceName {
+	return corev1.ResourceCPU
+}
+
+func (m *cpu) NodeResourceQuantity() resource.Quantity {
+	max := m.autoscalingSpec.CPURange.Max.MilliValue()
+	// Surface the situation where CPU is exhausted.
+	if m.requiredNodeCPUCapacity.MilliValue() > max {
+		// Elasticsearch requested more CPU per node than allowed by the user
+		err := fmt.Errorf("node required CPU is greater than the maximum one")
```
Comment (thbkrkr) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go:38]: I'm not sure but it sounds weird.
```suggestion
		err := fmt.Errorf("CPU required by the node is greater than the maximum allowed")
```

File: pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+			"policy", m.autoscalingSpec.Name,
+			"required_cpu", m.requiredNodeCPUCapacity,
+			"max_allowed_cpu", m.autoscalingSpec.CPURange.Max,
+		)
+		// Also update the autoscaling status accordingly
+		m.statusBuilder.
+			ForPolicy(m.autoscalingSpec.Name).
+			RecordEvent(
+				status.VerticalScalingLimitReached,
+				fmt.Sprintf("Node required CPU %s is greater than max allowed: %s", m.requiredNodeCPUCapacity, m.autoscalingSpec.CPURange.Max.String()),
```
Comment (thbkrkr) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go:51]:
```suggestion
				fmt.Sprintf("CPU required by the node %s is greater than the maximum allowed: %s", m.requiredNodeCPUCapacity, m.autoscalingSpec.CPURange.Max.String()),
```

File: pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+	log logr.Logger,
+	statusBuilder *status.AutoscalingStatusBuilder,
+	autoscalingSpec esv1.AutoscalingPolicySpec,
+	autoscalingPolicyResult client.AutoscalingPolicyResult,
+	currentAutoscalingStatus status.Status,
+) (Recommender, error) {
+	// Check if user expects the resource to be managed by the autoscaling controller
+	hasResourceRange := autoscalingSpec.CPURange != nil
+
+	// Did we get a resource requirement from Elasticsearch ?
```
Comment (thbkrkr) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go:105]:
```suggestion
	// Did we get a resource requirement from Elasticsearch?
```

File: pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go (hunk 3 above)
Comment (thbkrkr) [reply pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go:38]: required or requested?

File: pkg/controller/autoscaling/elasticsearch/testdata/cpu-scaled-horizontally/elasticsearch-expected.yml [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        - ReadWriteOnce
+        resources:
+          requests:
+            storage: 4Gi
+  version: 8.4.0
+status:
+  availableNodes: 10
+  health: green
+  phase: Ready
+  version: 7.11.0
```
Comment (thbkrkr) [inline pkg/controller/autoscaling/elasticsearch/testdata/cpu-scaled-horizontally/elasticsearch-expected.yml:82]:
```suggestion
  version: 8.4.0
```

**Review by thbkrkr — APPROVED** [review]
I didn't test in real but code LGTM.
Left some nitpicking on wording.

File: pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go (hunk 3 above)
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/cpu.go:38]: > required or requested?

required: https://www.elastic.co/guide/en/elasticsearch/reference/current/autoscaling-get-autoscaling-capacity.html#autoscaling-get-autoscaling-capacity-api-response-body

**Pull request merged**